PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -482,18 +603,105 @@ bool ProjectPanel::buildTreeFrom(TiXmlNode *projectRoot, HTREEITEM hParentItem)
 					return false;
 			}
 		}
+		if (lstrcmp(TEXT("Directory"), v) == 0)
+		{
+
+			const TCHAR *strLabel = (childNode->ToElement())->Attribute(TEXT("userLabel"));
+			generic_string newFolderLabel;
+			
+			const TCHAR *strValue = (childNode->ToElement())->Attribute(TEXT("name"));
+			generic_string fullPath = getAbsoluteFilePath(strValue);
+
+			std::vector<generic_string> filters;
+			const TCHAR *strFilters = (childNode->ToElement())->Attribute(TEXT("filters"));
+			if (strFilters)
+			{
+				generic_string filter(strFilters);
+				filters = stringSplit(filter, TEXT(";"), true);
+			}
+
+			if (strLabel)
+			{
+				newFolderLabel = strLabel;
+			}
+			else
+			{
+
+				newFolderLabel = buildDirectoryName(fullPath, filters);
+			}
+
+
+			HTREEITEM hNewDirectory = addDirectory(hParentItem, newFolderLabel.c_str(), true, fullPath.c_str(), &filters);
+			if (strLabel)
+			{
+				TVITEM tvItem;
+				tvItem.mask = TVIF_PARAM;
+				tvItem.hItem = hNewDirectory;
+				::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+				ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.